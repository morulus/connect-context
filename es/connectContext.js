/* eslint no-restricted-syntax: 0 */
import createContextConnector from './createContextConnector';
import defaultConnector from './defaultConnector';

export default createContextConnector(defaultConnector);