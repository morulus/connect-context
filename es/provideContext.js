function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/* eslint no-param-reassign: 0 */
import React, { Component } from 'react';
import invariant from 'invariant';
import mapStaticContextToTypes from './mapStaticContextToTypes';

export default function provideContext(contextProps, contextTypes) {
  var getChildContext = contextProps;
  var contextPropsType = typeof contextProps;
  var contextTypesType = typeof contextTypes;
  invariant(contextPropsType === 'function' || contextPropsType === 'object', 'contextProps must be a function or object, \'' + contextPropsType + '\' given;');
  invariant(contextPropsType !== 'function' || contextTypesType === 'object', 'If contextProps is function, contextTypes required');
  if (contextPropsType === 'object') {
    getChildContext = function getChildContext() {
      return contextProps;
    };
    if (!contextTypes) {
      contextTypes = mapStaticContextToTypes(contextProps);
    }
  } else {
    getChildContext = contextProps;
  }
  return function (Child) {
    var _class, _temp2;

    return _temp2 = _class = function (_Component) {
      _inherits(ContextProvider, _Component);

      function ContextProvider() {
        var _temp, _this, _ret;

        _classCallCheck(this, ContextProvider);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
          args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, _Component.call.apply(_Component, [this].concat(args))), _this), _this.getChildContext = getChildContext, _temp), _possibleConstructorReturn(_this, _ret);
      }

      ContextProvider.prototype.render = function render() {
        return React.createElement(Child, this.props);
      };

      return ContextProvider;
    }(Component), _class.childContextTypes = contextTypes, _temp2;
  };
}