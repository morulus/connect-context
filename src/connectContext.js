/* eslint no-restricted-syntax: 0 */
import React from 'react';
import { connect } from 'react-redux';
import invariant from 'invariant';

const PORT_CONTEXT_KEY = '@@connectContext.CONNECTOR_CONTEXT_TEMPORARY_CONTEXT_PROP';
const getHigherProxyComponent = (Child, contextTypes) => {
  const ProxyComponent = function ContextProxyComponent(props, context) {
    const nextProps = {
      [PORT_CONTEXT_KEY]: [context, props],
    };
    return React.createElement(Child, nextProps, nextProps.children);
  };

  ProxyComponent.contextTypes = contextTypes;
  return ProxyComponent;
};
/* eslint-disable react/prop-types */
const getLowerProxyComponent = Child => function ContextLowerProxyComponent(props) {
  const omittedProps = {};
  for (const prop in props) {
    if (
        Object.prototype.hasOwnProperty.call(props, prop)
      ) {
      if (prop !== PORT_CONTEXT_KEY) {
        omittedProps[prop] = props[prop];
      } else {
        Object.assign(omittedProps, props[prop][1]);
      }
    }
  }
  return React.createElement(Child, omittedProps, props.children);
};
/* eslint-enable */

export default function connectContext(contextTypes, mapContextToProps, ...others) {
  const typeOfContextTypes = typeof contextTypes;
  invariant(contextTypes !== 'object', `connectContext: first argument accepted as '${typeOfContextTypes}', expected an object, `);
  return (Child) => {
    const ProxyLower = getLowerProxyComponent(Child);
    const ConnectedToRedux = connect(
      (state, props) => mapContextToProps(
        props[PORT_CONTEXT_KEY][0],
        state,
        props[PORT_CONTEXT_KEY][1],
      ),
      ...others,
    )(ProxyLower);
    return getHigherProxyComponent(ConnectedToRedux, contextTypes);
  };
}
