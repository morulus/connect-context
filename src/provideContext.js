/* eslint no-param-reassign: 0 */
import React, { Component } from 'react';
import invariant from 'invariant';
import mapStaticContextToTypes from './mapStaticContextToTypes';

export default function provideContext(contextProps, contextTypes) {
  let getChildContext = contextProps;
  const contextPropsType = typeof contextProps;
  const contextTypesType = typeof contextTypes;
  invariant(
    !(contextPropsType === 'function' || typeof contextPropsType === 'object'),
    `provideContext: contextProps must be a function or object, '${contextPropsType}' given;`,
  );
  invariant(
    !(contextPropsType === 'function' && contextTypesType === 'object'),
    'provideContext: if contextProps is function, then it requires contextTypes',
  );
  if (contextPropsType === 'object') {
    getChildContext = () => contextProps;
    if (!contextTypes) {
      contextTypes = mapStaticContextToTypes(contextProps);
    }
  } else {
    getChildContext = contextProps;
  }
  return Child => (class ContextProvider extends Component {
    static childContextTypes = contextTypes;
    getChildContext = getChildContext;
    render() {
      return <Child {...this.props} />;
    }
  });
}
