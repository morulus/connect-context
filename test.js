import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import PropTypes from 'prop-types';
import provideContext from './es/provideContext';
import connectContext from './es/connectContext';

function noContextTypes() {
  const Child = connectContext({
    message: PropTypes.string.isRequired,
  }, ({ message }) => ({
    contextMessage: message,
  }))(({ contextMessage, ownPropMessage }) => <div>{contextMessage},{ownPropMessage}</div>);
  const message = 'Hello, Child!';
  const Provider = provideContext(() => ({
    message,
  }))(({ children }) => children);
  shallow(<Provider>
    <Child
      ownPropMessage="Hello, Provider!"
    />
  </Provider>);
}

describe('connectContext', () => {
  it('Provide & connect, mixin up with own prop', () => {
    const Child = connectContext({
      message: PropTypes.string.isRequired,
    }, ({ message }) => ({
      contextMessage: message,
    }))(({ contextMessage, ownPropMessage }) => <div>{contextMessage},{ownPropMessage}</div>);
    const message = 'Hello, Child!';
    const Provider = provideContext({
      message,
    })(({ children }) => children);
    expect(toJson(mount(<Provider>
      <Child
        ownPropMessage="Hello, Provider!"
      />
    </Provider>))).toMatchSnapshot();
  });

  it('Type/args errors', () => {
    expect(noContextTypes).toThrowError('If contextProps is function, contextTypes required');
  });
});
