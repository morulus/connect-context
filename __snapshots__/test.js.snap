// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectContext Provide & connect, mixin up with own prop 1`] = `
<ContextProvider>
  <Component>
    <ContextProxyComponent
      ownPropMessage="Hello, Provider!"
    >
      <ContextConnector
        @@connectContext.CONNECTOR_CONTEXT_TEMPORARY_CONTEXT_PROP={
          Array [
            Object {
              "message": "Hello, Child!",
            },
            Object {
              "ownPropMessage": "Hello, Provider!",
            },
          ]
        }
      >
        <ContextLowerProxyComponent
          @@connectContext.CONNECTOR_CONTEXT_TEMPORARY_CONTEXT_PROP={
            Array [
              Object {
                "message": "Hello, Child!",
              },
              Object {
                "ownPropMessage": "Hello, Provider!",
              },
            ]
          }
          contextMessage="Hello, Child!"
        >
          <Component
            contextMessage="Hello, Child!"
            ownPropMessage="Hello, Provider!"
          >
            <div>
              Hello, Child!
              ,
              Hello, Provider!
            </div>
          </Component>
        </ContextLowerProxyComponent>
      </ContextConnector>
    </ContextProxyComponent>
  </Component>
</ContextProvider>
`;
